File: Data Structures & Algorithms/trapping-rain-water/submission-0.py
class Solution:
    def trap(self, height: List[int]) -> int:

        totalWaterStore = 0
        maxLeft = []
        maxRight = []

        maxLH = 0
        for h in height:
            if h > maxLH:
                maxLH = h
            maxLeft.append(maxLH)

        maxRH = 0
        for h in reversed(height):
            if h > maxRH:
                maxRH = h
            maxRight.append(maxRH)
        maxRight.reverse()


        for i in range(len(height)):
            waterStored = min(maxLeft[i],maxRight[i]) - height[i]
            if waterStored > 0:
                totalWaterStore +=waterStored

        
        return totalWaterStore

        


        

File: Data Structures & Algorithms/trapping-rain-water/submission-0.java
class Solution {
    public int trap(int[] height) {

        int[] maxLeft = new int[height.length];
        int[] maxRight = new int[height.length];
        int total = 0;
        
        int currMax = 0;
        for (int i = 0; i < height.length; i++) {
            maxLeft[i] = currMax;
            currMax = Math.max(currMax, height[i]);
        }

        currMax = 0;
        for (int i = height.length-1; i >= 0; i--) {
            maxRight[i] = currMax;
            currMax = Math.max(currMax, height[i]);
        }

        for (int i = 0; i < height.length; i++) {
            int totalCap = Math.min(maxLeft[i], maxRight[i]) - height[i];
            if (totalCap > 0) total+= totalCap;
        }
        
        return total;
        
    }
}
